Document YOLACT support (#4749)
openvinotoolkit/openvino: docs/MO_DG/prepare_model/convert_model/onnx_specific/Convert_YOLACT.md

 
+            pred_outs.pop('priors') # remove unused in postprocessing layers
+            pred_outs.pop('loc') # remove unused in postprocessing layers
+            return pred_outs
 
 
 
-- 
```
3. Save and close the file.

## Convert YOLACT Model to the IR

**Step 1**. Clone the GitHub repository and check out the commit:

1. Clone the YOLACT reposirory:
```sh
git clone https://github.com/dbolya/yolact
```
2. Check out the necessary commit:
```sh
git checkout 57b8f2d95e62e2e649b382f516ab41f949b57239
```

**Step 2**. Download pre-trained model, for example `yolact_base_54_800000.pth`.

**Step 3**. Export model to ONNX* format.

1. Apply the `YOLACT_onnx_export.patch` patch to the repository. Refer to the <a href="#patch-file">Create a Patch File</a> instructions if you do not have it:
```sh
 git apply /path/to/patch/YOLACT_onnx_export.patch
```

2. Evaluate the YOLACT mode in order to export to ONNX* format:

```sh
python3 eval.py
    --trained_model=/path/to/yolact_base_54_800000.pth
    --score_threshold=0.3
    --top_k=10
    --image=/path/to/image.jpg
```

3. You should get `yolact.onnx` file.

**Step 4**. Convert the model to the IR:

```sh

> python path/to/model_optimizer/mo.py --input_model /path/to/yolact.onnx

Reviewer: what about mean, scale and reverse_input_channles.
In OMz we provide following arguments for conversion:
https://github.com/openvinotoolkit/open_model_zoo/blob/master/models/public/yolact-resnet50-fpn-pytorch/model.yml#L99-L104

But they are model specific (resnet50/resnet101 is the same, darknet53 different).
Mean and scale values can be obtained from model config https://github.com/dbolya/yolact/blob/master/data/config.py

Author: We generally don't provide such arguments for model conversion documentation. @lazarevevgeny what is your impression on this issue?

Reviewer: as I see they are provided in YOLO, but if it is not so important than ok

Reviewer 2: Regarding comment that we don't have such arguments in documents. In these cases they are probably not needed because the mean/scale are a part of the model itself. My opinion is that we should provide mean/scale values so the customer has full understanding what is going on and how to infer the model.

# Convert PyTorch* YOLACT to the Intermediate Representation {#openvino_docs_MO_DG_prepare_model_convert_model_onnx_specific_Convert_YOLACT}

You Only Look At CoefficienTs (YOLACT) is a simple, fully convolutional model for real-time instance segmentation.

> The PyTorch* implementation is publicly available in [this GitHub* repository](https://github.com/dbolya/yolact).

Reviewer 3: ```suggestion
The PyTorch\* implementation is publicly available in [this GitHub* repository](https://github.com/dbolya/yolact).
```

The YOLACT++ model is not supported, because it uses deformable convolutional layers that cannot be represented in ONNX* format.

## Create a Patch File <a name="patch-file"></a>

> Before converting the model, you need to create a patch file for the repository.

Reviewer 3: ```suggestion
Before converting the model, create a patch file for the repository.
```

     
diff --git a/utils/augmentations.py b/utils/augmentations.py
index cc7a73a..2420603 100644
--- a/utils/augmentations.py
+++ b/utils/augmentations.py
@@ -623,8 +623,11 @@ class FastBaseTransform(torch.nn.Module):
     def __init__(self):
         super().__init__()
 
-        self.mean = torch.Tensor(MEANS).float().cuda()[None, :, None, None]
-        self.std  = torch.Tensor( STD ).float().cuda()[None, :, None, None]
+        self.mean = torch.Tensor(MEANS).float()[None, :, None, None]
+        self.std  = torch.Tensor( STD ).float()[None, :, None, None]
+        if torch.cuda.is_available():
+            self.mean.cuda()
+            self.std.cuda()
         self.transform = cfg.backbone.transform
 
     def forward(self, img):
diff --git a/yolact.py b/yolact.py
index d83703b..f8c787c 100644
--- a/yolact.py
+++ b/yolact.py
@@ -17,19 +17,22 @@ import torch.backends.cudnn as cudnn
 from utils import timer
 from utils.functions import MovingAverage, make_net
 
-# This is required for Pytorch 1.0.1 on Windows to initialize Cuda on some driver versions.
-# See the bug report here: https://github.com/pytorch/pytorch/issues/17108
-torch.cuda.current_device()
-

> -# As of March 10, 2019, Pytorch DataParallel still doesn't support JIT Script Modules

Reviewer 3: ```suggestion
-# As of March 10, 2019, Pytorch DataParallel still does not support JIT Script Modules
```

Author: This text is part of the patch file, should not be changed, since it is in the original repository.

-use_jit = torch.cuda.device_count() <= 1
-if not use_jit:

> -    print('Multiple GPUs detected! Turning off JIT.')

Reviewer 3: ```suggestion
-    print('Multiple GPUs detected. Turning off JIT.')
```

+use_jit = False
 
 ScriptModuleWrapper = torch.jit.ScriptModule if use_jit else nn.Module
 script_method_wrapper = torch.jit.script_method if use_jit else lambda fn, _rcn=None: fn
 
 
+def decode(loc, priors):
+    variances = [0.1, 0.2]
+    boxes = torch.cat((priors[:, :2] + loc[:, :, :2] * variances[0] * priors[:, 2:], priors[:, 2:] * torch.exp(loc[:, :, 2:] * variances[1])), 2)
+
+    boxes_result1 = boxes[:, :, :2] - boxes[:, :, 2:] / 2
+    boxes_result2 = boxes[:, :, 2:] + boxes[:, :, :2]
+    boxes_result = torch.cat((boxes_result1, boxes_result2), 2)
+
+    return boxes_result
+
 
 class Concat(nn.Module):
     def __init__(self, nets, extra_params):
@@ -476,7 +479,10 @@ class Yolact(nn.Module):
     
     def load_weights(self, path):
         """ Loads weights from a compressed save file. """
-        state_dict = torch.load(path)
+        if torch.cuda.is_available():
+            state_dict = torch.load(path)
+        else:
+            state_dict = torch.load(path, map_location=torch.device('cpu'))
 

>          # For backward compatability, remove these (the new variable is called layers)

Reviewer 3: ```suggestion
         # For backward compatibility, remove these (the new variable is called layers)
```

Author: Same as above, patch file.
